[RDY] Don't allow changing encoding after startup scripts

bfredl opened the pull request:
In #2469 @justinmk suggested to not allow changing `&encoding` after `VimEnter`. This PR actually forbids it slightly earlier, after nvimrc is sourced, but before plugins are loaded. The rationale is that only the user should be allowed to override the locale encoding detection, and this way any plugin could rely on encoding being unchanged after it's `{rtp}/plugin/` file is sourced, it could do once:

```
if &encoding != "utf-8"
     echomsg "warning: nvim-myfancyplugin is only fully functional when encoding=utf-8"
endif
```

or an rplugin only needs to check `vim.options['encoding']` (and setup conversion) once, IIRC some /n/vim plugins already assume this,  but with this change it is completely reliable.

Also cleanup the TUI encoding initialization, should probably be squashed but separated it now for easier review.

bfredl committed:
encoding: simplify handling of encoding in TUI

bfredl commented:
The error message prehaps should be made generic, I could imagine we want to do the same for options. What is our policy for allocating `Exxx` numbers?

bfredl commented:
should probably add `encoding=utf-8` to the `nvim_argv` in helpers and remove it from the `095_regexp_multibyte_spec`

Pyrohh commented:
> should probably add encoding=utf-8 to the nvim_argv in helpers and remove it from the 095_regexp_multibyte_spec

Do you even need to set it explicitly? I thought that's already the default of `encoding` (assuming a UTF-8 locale is set via the environment).

justinmk commented on src/nvim/globals.h:
```diff
@@ -746,6 +746,9 @@ EXTERN bool enc_utf8 INIT(= false);             /* UTF-8 encoded Unicode */
 EXTERN int has_mbyte INIT(= 0);                 /* any multi-byte encoding */
 
+// 'encoding' can only be changed in the startup scripts
+EXTERN bool enc_mutable INIT(= true);
```
Is there a variable that already tracks whether `VimEnter` was raised? Or, can we name this something more generalized, which describes the phase that has been completed, e.g. `is_init_completed`.

bfredl replied:
The problem is that "at VimEnter" might be to late: both BufEnter and VimEnter are executed _after_ the TUI is initialized, yes it is a contrived example, but then it is possible to do
`au BufEnter * set encoding=something`
and possibly break the TUI input.
Why not just rename `enc_mutable` to `startupscripts_loaded` or something?

justinmk replied:
> Why not just rename enc_mutable to startupscripts_loaded or something

Ok. The convention in globals.h seems to be `did_foo`, so how about `did_source_startup_scripts` and set it at the end of `source_startup_scripts()`.

I was wrong about `current_SID` by the way, it is 0 unless a script is running.

justinmk commented:
Note to self: At `:help 'encoding'`, in addition to mentioning that it is not settable after VimEnter, we should direct users to use 'fileencoding', `e ++enc` (as well as vim's `icon()` function) to set per-file encoding, since `'encoding'` is so often misunderstood.

bfredl commented:
@Pyrohh The point was to allow running `make test` even if the locale makes `encoding=latin-1`. Now, I checked, there are several other functional tests that already break if `LC_ALL=en_US make test`, but they all  work if we add `encoding=utf8` to `helpers.lua`.

Somewhat unrelated, oldtests/test87 is a guaranteed failure locally for anyone that `has("python3")` as it tests `bindeval`.

bfredl commented:
I will disable test87 the same way as test86 (python2) was

bfredl committed:
encoding: Update handling of encoding in tests

Always run tests with encoding=utf-8, regardless of user locale
Don't set &encoding after startup in tests

Helped-By: Michael Reed <m.reed@mykolab.com>

justinmk commented on src/nvim/testdir/test87.in:
```diff
@@ -3,7 +3,7 @@ Tests for various python features.     vim: set ft=vim :
 :so small.vim
 :set noswapfile
-:if !has('python3') | e! test.ok | wq! test.out | endif
+:if !has('python3') || has('nvim') | e! test.ok | wq! test.out | endif
```
test86 and test87 should just be deleted. The python client has its own test suite. cc @tarruda

bfredl replied:
seems was fixed in #2650 just merged

tarruda replied:
> test86 and test87 should just be deleted. The python client has its own test suite. cc @tarruda

:+1: no need to keep these around

justinmk commented on src/nvim/option.c:
```diff
@@ -3696,7 +3696,9 @@ did_set_string_option (
   else if (varp == &p_enc || gvarp == &p_fenc) {
-    if (gvarp == &p_fenc) {
+    if (varp == &p_enc && !enc_mutable) {
+       errmsg = (char_u *)N_("Cannot be changed after initialization");
```
We need a new EXX error number for this, as well as a blurb in the docs.

bfredl replied:
What is the policy for allocating a number? does neovim have its own series ?

bfredl commented:
Updated. Also need to fix oldtests 39, 58, 59 (they didn't fail locally last time for some reason...)

bfredl commented:
39 was simple, 58/59 seems trickier, especially 59 which combines latin-1 and utf-8 in the same file. (with the motivation: "Don't want to depend on the locale from the environment")

Pyrohh commented:
I merged #3074, so you can drop that commit now (well, whenever you get around to it).

bfredl commented:
Rebased. Needs converted test 59 #3076 . I would suggest dropping 58; it is identical to 59 except that the test text is latin-1 encoded. Test 52 is weird, for some reason it is marked to only run on windows and thus has been effectively disabled, so I won't regard it as a blocker for this PR, but later on its spirit could carry on in a test that does the same transformations with `++enc` instead of changing `&enc`.

justinmk commented:
Note to self: need to test `set all&` called after VimEnter and `'enc'` was set to something other than the default.

bfredl commented:
`:set enc=somethingnotdefault` and then `:set all&` is broken already on master. (The text value is changed, but `mb_init` is not called). On this branch the fix is easy: make &encoding `P_NODEFAULT`

justinmk commented:
> On this branch the fix is easy

Awesome, that has been bugging me for awhile. That makes this PR extra valuable :)

If I understand correctly, https://github.com/neovim/neovim/pull/3076 is blocking this PR, so my plan was to finish that up (review it, etc) before this.

justinmk commented:
Can you rebase this so that the test58/59 changes factored out?

bfredl commented:
I have remove test58/59 temporarily to be able to interpret travis results at all (otherwise I hit travis 4MB log limit, yes it's true) I will of course remove that commit once #3076, any reason you want it removed now?

justinmk commented:
Oh, I'm sorry, I thought those were the python3 tests.

bfredl commented:
No problem :) BTW `set enc` + `:set all&` is bork in vim as well, at least latest arch packaged version.

bfredl commented:
oops, it broke `LC_ALL=C nvim` (but not `LC_ALL=en_US` ), perhaps should only set `P_NODEFAULT` after initialization.

bfredl commented:
Posted `set all&` issue to [vim-dev](https://groups.google.com/forum/#!topic/vim_dev/0UUibTqIBj0)

bfredl commented:
Related vim-patch vim/vim@b341dda5 , though the `set all&` issue turned out to more general (same vim-dev thread).

bfredl commented:
rebased

Pyrohh commented on runtime/doc/options.txt:
```diff
@@ -2219,10 +2219,10 @@ A jump table for the options with a short description can be found at |Q_op|.
-	It should normally be kept at its default value, or set when Vim
-	starts up.  See |multibyte|.  To reload the menus see |:menutrans|.
+	This option cannot be changed after sourcing of |vimrc| files are done.
+	Otherwise it would have caused non-ASCII text inside Vim to become
```
maybe just `otherwise it would cause`

Pyrohh replied:
Also I think you can remove the whole paragraph regarding GTK+ 2 (just below the modeline one).

Pyrohh replied:
Or not, I forgot this wasn't the PR which makes utf-8 the only option. Sorry.

bfredl replied:
Following the discussion in #1250 we should update the GTK paragraph to a more general statement about GUI:s and remote plugins might be only fully functional when `encoding=utf-8` (at least not  "care has been taken to allow different values of
	'encoding'" )

Pyrohh commented:
Should probably be mentioned in `vim_diff.txt`.

bfredl commented:
I did some update to the docs, but not too much, as the situation will change again with #2905 .

bfredl committed:
encoding: don't allow changing encoding after startup scripts

bfredl committed:
encoding: update docs for encoding

Helped-By: Michael Reed <m.reed@mykolab.com>
Helped-By: Justin M. Keyes <justinkz@gmail.com>

bfredl committed:
encoding: test that `&encoding` cannot be changed

Helped-By: Justin M. Keyes <justinkz@gmail.com>

bfredl commented:
Also added a test. (no category really fit but in the end I placed it in `ex_cmds` as it tests behaviour of `:set``)

justinmk commented:
I would like to take a careful look at this tonight, so please wait. Sorry for the delay.

bfredl commented:
:+1: No problem!

Pyrohh commented on runtime/doc/vim_diff.txt:
```diff
@@ -172,6 +172,8 @@ Other compile-time features:
   EBCDIC
   Emacs tags support
 
+'encoding' cannot be changed after sourcing of |vimrc| files is done.
```
I think this should be under the "Changed Features" section instead.

bfredl replied:
:+1:

Pyrohh commented on runtime/doc/mbyte.txt:
```diff
@@ -89,8 +89,8 @@ See |encoding-values| for a list of acceptable values.
 encoding.  Not only the text in the buffers, but also in registers, variables,
-etc.  This also means that changing the value of 'encoding' makes the existing
-text invalid!  The text doesn't change, but it will be displayed wrong.
+etc. 'encoding' cannot be changed after initialization, as changing the value
```
if "initialization" corresponds to that documented under "3. Initialization" in starting.txt, perhaps it should be linked to.

bfredl replied:
more specifically point 3  ("vimrc") underneath that. The point in this paragraph was more about open buffers shouldn't be corrupted, rather than precisely when `&encoding` gets frozen, but certainly it won't hurt to add a link to the relevant place :+1:

Pyrohh commented:
I see you've removed some explicit setting of `encoding`; perhaps this could be done as well:
```diff
diff --git a/test/functional/legacy/082_string_comparison_spec.lua b/test/functional/legacy/082_string_comparison_spec.lua
index bd4e8a4..1615828 100644
--- a/test/functional/legacy/082_string_comparison_spec.lua
+++ b/test/functional/legacy/082_string_comparison_spec.lua
@@ -9,7 +9,6 @@ describe('case-insensitive string comparison in UTF-8', function()
   setup(clear)
 
   it('is working', function()
-    execute('set enc=utf8')
     feed('ggdG<cr>')
     source([[
       function! Ch(a, op, b, expected)

```

bfredl commented:
:+1: missed that one

bfredl commented:
@Pyrohh Thanks, updated. Also removed a few more superfluous `set enc[oding]=utf-8`

bfredl commented:
Updated for #3246. @justinmk did you take a look?

justinmk commented:
Not yet, I am traveling.

justinmk commented on runtime/doc/message.txt:
```diff
@@ -739,6 +739,12 @@ Example: >
 a user-defined command.
 
+							*E905*  >
+  Cannot be changed after nvimrc files are sourced
```
"nvimrc" is probably going away after XDG is adopted. Suggestion:

    Cannot set this option after startup.

"startup" is used in several areas of the Vim docs and is probably the most accurate and recognizable term.

bfredl replied:
Technically it is less accurate, |vimrc| (forgot the link here) refers to step 3 in |startup|, but as plugins have no business changing `encoding`, "after startup" should be clear enough. :+1:

bfredl replied:
nevermind, I mention |vimrc| on line below...

justinmk commented on runtime/doc/mbyte.txt:
```diff
@@ -81,7 +81,7 @@ See |mbyte-locale| for details.
 
 If your locale works properly, Vim will try to set the 'encoding' option
-accordingly.  If this doesn't work you can overrule its value: >
+accordingly.  If this doesn't work you can overrule its value in |vimrc|: >
```
Might as well forget mentioning "vimrc".

bfredl replied:
:+1:

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2217,21 +2217,16 @@ A jump table for the options with a short description can be found at |Q_op|.
-	the default to prevent different behavior of the GUI and terminal
-	versions, and to avoid changing the encoding of newly created files
-	without your knowledge (in case 'fileencodings' is empty).
+	This option cannot be changed after sourcing of |vimrc| files is done.
```
This option cannot be changed after startup.

bfredl replied:
:+1:

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2217,21 +2217,16 @@ A jump table for the options with a short description can be found at |Q_op|.
+	This option cannot be changed after sourcing of |vimrc| files is done.
+	Otherwise it would cause non-ASCII text inside Vim to become
+	invalid. It should normally be kept at its default value, or be set
+	in vimrc. See |multibyte|. To reload the menus see |:menutrans|.
```
`:menutrans` doesn't seem relevant here now.

bfredl replied:
:+1:

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2217,21 +2217,16 @@ A jump table for the options with a short description can be found at |Q_op|.
+	recommended to set 'encoding' to "utf-8". Plugins are free to
+	assume that 'encoding' is "utf-8" and might only support non-ASCII
+	text in this case. On most modern platforms the default value will
+	already be "utf-8".
```
Paragraph can be replaced with:

    The recommended 'encoding' is "utf-8". Remote plugins only support utf-8.

bfredl replied:
:+1:

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -74,6 +74,8 @@ are always available and may be used simultaneously in separate plugins.  The
    'p')) mkdir() will silently exit. In Vim this was an error.
 3. mkdir() error messages now include strerror() text when mkdir fails.
 
+'encoding' cannot be changed after sourcing of |vimrc| files is done.
```
'encoding' cannot be changed after startup.

bfredl replied:
:+1:

justinmk commented on test/functional/ex_cmds/encoding_spec.lua:
```diff
@@ -0,0 +1,39 @@
+    execute('set encoding=latin1')
+    -- error message expected
+    feed('<cr>')
+    eq('utf-8', eval('&encoding'))
```
could also check:

    neq(nil, string.find(eval('v:errmsg'), '^E905: '))

bfredl replied:
nice!

justinmk commented:
Done reviewing..

bfredl commented:
Thanks, updated.
